wordpress-mobile/release-toolkit
# Interactive Prompt Reminder for lanes waiting for input/confirm

## What

This introduces a [monkey-patch](https://en.wikipedia.org/wiki/Monkey_patch) on _fastlane_'s interactive UI methods like the `UI.input` and `UI.confirm` which will remind you (via beep + badge and icon jumping in Dock, and optionally an opt-in spoken alert) that the Terminal is waiting for an answer from you if you didn't provide one after some time.

## Why

This is particularly very useful if, like me, your often run a command like `bundle exec fastlane new_beta_release` then switch another app to work on something else, thinking that you can work on something else while the beta is building on CI… only to realize 10mn or even 30mn or more later that the CI never started, and that you've wasted all that waiting time for nothing, because you forgot that the lane was waiting for you to confirm the action first.

With this change, if you go do other things and forget that the prompt is waiting for you, you'll now be reminded about it ✨ 

## Demo

_(Turn sound on, as I narrate and have enabled the `FASTLANE_PROMPT_REMINDER_MESSAGE` option too)_

https://user-images.githubusercontent.com/216089/137374406-4e6d4985-d8a2-4c35-a296-0b409915d75c.mov

## Customization

This code uses Monkey-Patching 🐒  to redefine _fastlane_'s `UI.…` interactive methods (`UI.input`, `UI,confirm`, etc) and wrap their original implementations around a helper method (`UI.with_reminder`) which handles the delay+alerting logic.

By default, those `UI.…` methods will be automatically patched to generate a system beep and make your Terminal icon jump after 30s then 3mn then 10mn. You can opt-out of the auto-patching using `FASTLANE_PROMPT_REMINDER_DISABLE_AUTO_PATCH=1` env var.

When you _don't_ opt-out of the auto-patching (the default), you can also refine the behavior of those reminders:
 - By default, it will only beep + make the Dock icon bounce, after 30s, 180s then 600s of you failing to reply the prompts
 - You can customize the delays using `FASTLANE_PROMPT_REMINDER_DELAYS` env var
 - You can also make the reminder **speak** a message – via the System's `say` command – using the `FASTLANE_PROMPT_REMINDER_MESSAGE` env var.
    - Set it to the value `default`, `yes`, `true` or `1` to use the default message, i.e "An interactive prompt is waiting for you in the Terminal!"
    - Set it to any other string to use that string as the message to be spoken
    - Unset it if you don't want the reminder to speak to you, but only beep + bounce dock icon (the default)

### To test

<details><summary>0. Point your Pluginfile to this PR</summary>

   - Cut a test branch from your `develop` branch in your client repo (e.g. WPiOS)
   - Point your `fastlane/Pluginfile` to this branch, run `bundle update fastlane-plugin-wpmreleasetoolkit` to update the `Gemfile.lock`, and commit the changes (because we need a clean repo in order for the `new_beta_release` lane not to fail)

</details>

<details><summary>1. Test the default behavior</summary>

 - Run `bundle exec fastlane new_beta_release` in your terminal, then switch to another app and/or go drink a coffee (without waiting or replying to the prompt)
 - Confirm that after 30s while not replying to the prompt waiting for you in the terminal, you'll get a beep and the Terminal icon will jumping (once) in your Dock
 - Reply `n` to the prompt to cancel the new beta
 - `git checkout` back to your test branch (as `new_beta_release` would have switched to the release branch during its exececution)

</details>

<details><summary>2. Test a custom delay + default spoken message</summary>

 - Run `FASTLANE_PROMPT_REMINDER_DELAYS=3,5 FASTLANE_PROMPT_REMINDER_MESSAGE=default bundle exec fastlane new_beta_release`
 - Confirm that after 3 seconds, you will not only get the system beep + Dock icon badge and bounce, but also have your Mac speak out loud the "An interactive prompt is waiting for you in the Terminal!" message.
 - Confirm that after 5 more seconds, you'll get the beep + icon badge + spoken message once more
 - Reply `n` to the prompt, then `git checkout` back to your test branch

</details>

<details><summary>3. Test a custom spoken message</summary>

 - Run `FASTLANE_PROMPT_REMINDER_DELAYS=3,5 FASTLANE_PROMPT_REMINDER_MESSAGE="Hello?" bundle exec fastlane new_beta_release`
 - Confirm that after 3 seconds, you will not only get the system beep + Dock icon badge and bounce, but also have your Mac speak out loud the "An interactive prompt is waiting for you in the Terminal!" message.
 - Confirm that after 5 more seconds, you'll get the beep + icon badge + spoken message once more
 - Reply `n` to the prompt, then `git checkout` back to your test branch

</details>

<details><summary>4. Test opting out completely</summary>

 - Run `FASTLANE_PROMPT_REMINDER_DISABLE_AUTO_PATCH=1 FASTLANE_PROMPT_REMINDER_DELAYS=3,5 FASTLANE_PROMPT_REMINDER_MESSAGE=default bundle exec fastlane new_beta_release`
 - Confirm that you won't get any reminder at all, neither after 3, 5 or 30 seconds of waiting.
 - Reply `n` to the prompt, then `git checkout` back to your test branch

</details>

### Mergeability

We maybe want to wait for us to do the `2.0.0` major release of the `release-toolkit` first before merging this and ship it as part of a `2.1.0`, just because if we need to fix or revert things introduced by the breaking changes from #300 due to some bug, it would make more sense to keep the major release focused on the big change only. This is why I've kept this PR as Draft for now, and why I didn't add the `CHANGELOG` entry just yet.

That being said, if it's approved in time, and you feel it has its place in the upcoming `2.0.0`, I don't have strong feelings about it being included sooner.
